File: src/Sharp-O-Matic.Engine/Connections/FieldDescriptor.cs

namespace SharpOMatic.Engine.Connections;

public class FieldDescriptor
{
    public required string Name { get; set; }
    public required string Label { get; set; }
    public required string Description { get; set; }
    public required FieldDescriptorType Type { get; set; }
    public required bool IsRequired { get; set; }
    public required object? DefaultValue { get; set; }
    public required List<string> EnumOptions { get; set; }
}

File: src/Sharp-O-Matic.Editor/src/app/metadata/definitions/field-descriptor.ts
import { FieldDescriptorType } from '../enumerations/field-descriptor-type';

export interface FieldDescriptorSnapshot {
  name: string;
  label: string;
  description: string;
  type: FieldDescriptorType;
  isRequired: boolean;
  defaultValue: unknown;
  enumOptions: string[];
}

export class FieldDescriptor {
  constructor(
    public readonly name: string,
    public readonly label: string,
    public readonly description: string,
    public readonly type: FieldDescriptorType,
    public readonly isRequired: boolean,
    public readonly defaultValue: unknown,
    public readonly enumOptions: string[],
  ) {}

  public static fromSnapshot(snapshot: FieldDescriptorSnapshot): FieldDescriptor {
    return new FieldDescriptor(
      snapshot.name,
      snapshot.label,
      snapshot.description,
      snapshot.type,
      snapshot.isRequired,
      snapshot.defaultValue,
      [...snapshot.enumOptions],
    );
  }
}